server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -1904,10 +1904,21 @@ public void removeCustomOfferingDetails(long vmId) {
     }
 
     @Override
-    public HashMap<String, VolumeStatsEntry> getVolumeStatistics(long clusterId, String poolUuid, StoragePoolType poolType, List<String> volumeLocator, int timeout) {
+    public HashMap<String, VolumeStatsEntry> getVolumeStatistics(long clusterId, String poolUuid, StoragePoolType poolType, List<String> volumeLocators, int timeout) {
         List<HostVO> neighbors = _resourceMgr.listHostsInClusterByStatus(clusterId, Status.Up);
         for (HostVO neighbor : neighbors) {
-            GetVolumeStatsCommand cmd = new GetVolumeStatsCommand(poolType, poolUuid, volumeLocator);
+            StoragePool storagePool = _storagePoolDao.findPoolByUUID(poolUuid);
+            List<String> volumeLocatorsForHost;
+            if (storagePool.isManaged()) {
+                List<UserVmVO> vmsPerHost = _vmDao.listByHostId(neighbor.getId());
```

Comment:
looks good but can you isolate the body of the if in a method like filter hostsToQueryForVmVolumes() (and maybe add a test)? Please don't kill me but it is kind of hard to read like this.